Add variants required for square blue icon share links

AshGDS opened the pull request:
## What / Why
<!-- What are the reasons behind this change being made? -->
- Add variants required for square blue icon share links
- Reasoning: https://trello.com/c/LNOqGMgZ/49-build-social-media-links
- The existing component styles use `position: absolute` for the icons, which doesn't work nicely when we need them square. This is because the absolutely positioned icons don't affect the width/height of the DOM, so adding square background styles just ended up in overlapping, fiddly overflows, and made it difficult to align the text in the centre of the icons.
- Therefore, I've used `display: flex` along with `inline-block` to get the styles we want.
- ~~I added a `min-width` at a certain breakpoint to make the reflow look nicer, but this is specific to the page we're going to put this variation on. (i.e. because we know the social icons / share link text that will be on the page, we can make the shorter sized social links look nicer when they stack next to longer social links by adding a `min-width`.) Of course that CSS rule is not very helpful if this variation was used on other pages that are using an arbitrary number of social links.~~

## Visual Changes
<!-- If change results in visual changes, include detailed screenshots that show the various states. -->

<!-- Please ensure that the changes are reviewed by a Designer if required. -->
<!-- To help Designers, please include a link to specific elements to review, -->
<!-- for example to https://components-gem-pr-[PULL REQUEST NUMBER].herokuapp.com/public -->

Multiple examples added, so will need a Percy review.

<img width="997" alt="image" src="https://github.com/user-attachments/assets/65780164-1c71-485d-8a3a-217ac08777ae">

<img width="997" alt="image" src="https://github.com/user-attachments/assets/640d5265-d4ca-40da-8090-1f9b348b7815">

<img width="997" alt="image" src="https://github.com/user-attachments/assets/53eca06d-998b-46f2-be72-4ef8ad558f7d">

<img width="997" alt="image" src="https://github.com/user-attachments/assets/a9662d58-725f-4a47-8a1a-c21180841069">

<img width="1007" alt="image" src="https://github.com/user-attachments/assets/ca4483be-8e3e-4e71-a812-05fed7d2e283">

andysellick commented on app/assets/stylesheets/govuk_publishing_components/components/_share-links.scss:
I think this new option `square_blue_icons ` is trying to do two things at once (change the icons and change the layout) and I think it would be better to separate layout from appearance in this component.

Did you try using the `stacked` option? It seems to do almost exactly what's needed here in terms of layout - and then you could apply the blue icons as a separate option (allowing more flexibility by separating out the layout change from the icon change).

AshGDS replied:
Thanks for reviewing @andysellick - I'm not sure exactly what you mean here, as `stacked` would make the links vertical no matter what, but they need to be along one row for desktop. If you meant `columns: true`, I believe as that's based on an assumed width for the links it doesn't work as well here as this variation is a lot bigger - unless you want me to make a modification of that variation to try and work here?

Also, looking into it, `display: flex` wasn't actually needed - so I have removed that from the code.

I can remove the `min-width` rule if needed too, if that's what's causing concern, as it was a visual enhancement rather than something necessary. It would then look like this when stacked: 

## With min-width
<img width="549" alt="image" src="https://github.com/user-attachments/assets/a44cd838-c1c0-48d4-bfdb-57a57eaaca80">

## Without
<img width="549" alt="image" src="https://github.com/user-attachments/assets/18a1e12b-9ed3-49a5-9df4-8f94ab4546d8">

If you're referring to the icons themselves, do you mean to have a `relative_square_icons: true` variation, and then a colour option that would have the colours `dark blue / white` passed to it?

andysellick replied:
Sorry, I meant the `columns` option 🤦

AshGDS replied:
@andysellick No worries - I think using `columns` makes the spacing of the share links a bit less even, as it's using `display: grid` with a `min-width` to achieve this, so it makes some of the links wider than needed, which is noticeable on this variation. Also the `column` variation's `min-width` gets a bit tight for our share links when you resize the page (the icons nearly butt right up against each other before falling onto a new row,) so the CSS would probably need to be adjusted. I think this variation would work if the links/icons were smaller, but as they're larger it isn't as compatible. I think the squareness of the icons also shows the uneven spacing a bit clearer if we used this, than if we were using our standard icons. Therefore do you think it's still worth going with this sort of technique?

See the Flickr link on the examples for the spacing differences:

## With columns option

<img width="965" alt="image" src="https://github.com/user-attachments/assets/1b6ef201-dbd6-466b-80de-84a4567b6279">

## Without

<img width="965" alt="image" src="https://github.com/user-attachments/assets/ddb3eac1-d0e0-489b-a2ed-c8748fac249a">

andysellick replied:
The design is probably a bit open to interpretation but yeah they do seem to be variable width. I still think we should separate the layout from the presentation though. Is it worth doing one option for the new icons, and another one for the layout - maybe try some new layout with flexbox? Seems like a good candidate over grid.

![Screenshot 2024-10-09 at 15 45 32](https://github.com/user-attachments/assets/60051bab-e47c-4a64-8a18-42c3b4086e56)

matthillco replied:
Note that the icons look broken in Firefox with 'Zoom Only Text' option and the font size increased: 
<img width="400" alt="image" src="https://github.com/user-attachments/assets/229116fd-6fca-4017-a3d8-022e2bf73ff5">

It's easily fixed if you change the `display: inline-block` to `display: inline-flex` in `.gem-c-share-links__link-icon--square-blue`. The flexbox defaults will then auto-center the icon:

<img width="400" alt="image" src="https://github.com/user-attachments/assets/0afa32c9-6cb6-46b8-b5c5-19eaf5a55c96">

matthillco reviewed (commented):
I don't have anything to add to the discussion about the layout with Andy, the overall solution seems robust and rational to me. I did play with it a bit to see if we could remove the 180px min-width which is a bit of a magic number but it seems required: no combination of flex alignments/adjustments would solve it.

I just have one comment about the icons breaking in Firefox zoom text only mode. https://github.com/alphagov/govuk_publishing_components/pull/4292#discussion_r1794297771

AshGDS commented:
Thanks for the suggestions @andysellick - It should be ready for another review. I've refactored the implementation completely to try and separate layout and presentation. Hopefully I haven't overdone it, as I've ended up with five different variants. However I couldn't think of an other way of representing all the changes needed without creating the five variants. All of the new variants combine together to give us the icon styles we need for our other project. I've added them as separate commits to try and make it easier to review.

@matthillco Thanks for looking at the Firefox zoom as I didn't check that, I've incorporated your suggested fix into the work 👍 .

andysellick commented:
For the flexbox layout, I was wondering about this kind of thing...

```CSS
.gem-c-share-links--flexbox {
  .gem-c-share-links__list {
    display: flex;
    flex-wrap: wrap;

    @include govuk-media-query($until: tablet) {
      flex-direction: column;
    }
  }

  .gem-c-share-links__list-item {
    flex: 1 1 0px;
  }
}
```

Which gives the below - expands to fill the available width on a wide screen, but allows for overlap with extreme content, then collapsing to a single column on mobile.

Mobile | Tablet | Desktop 
------ | ------ | -----
![Screenshot 2024-10-10 at 15 55 15](https://github.com/user-attachments/assets/2ff17dcb-85b0-4985-a184-ed903180ada5) | ![Screenshot 2024-10-10 at 15 55 08](https://github.com/user-attachments/assets/42ebeb7f-2435-48f6-9a6e-bcc4a0194190) | ![Screenshot 2024-10-10 at 15 54 57](https://github.com/user-attachments/assets/f644521e-78d7-46da-b7c7-c30eab823374)

andysellick commented:
@AshGDS I think there might be too many options here now. I'd do one for the layout, and one for the new icon style.

AshGDS commented:
Thanks @andysellick - I have moved back to two variants and pushed a new commit to review.

Thanks for writing that code - on the square icons it behaves a bit differently because my icons were `inline-block` which meant they influenced the width of the links, and therefore led to scenarios like this: 
<img width="764" alt="image" src="https://github.com/user-attachments/assets/98618c8f-6f06-419b-8545-3b219fb136fd">

Therefore I've reverted back to `position: absolute` to try and better accommodate using your flex options. Using `position: absolute` which was simpler than I initially thought, apologies. This works well, but when two icons are on a row on their own it will look like the below screenshot - are we OK with this?:

<img width="687" alt="image" src="https://github.com/user-attachments/assets/ea880796-26f2-417d-bf0c-57d0beb79287">

As I'm back using `position: absolute` for the icons, we can also go back to just using the existing `columns: true` functionality, which makes it collapse like the non-square icons would. The minor flaw with `columns: true` is that the text will wrap even if there is still space for the text to the right.  

<img width="941" alt="image" src="https://github.com/user-attachments/assets/a68ab9c7-b040-47eb-b0f3-dbef586f16fe">

I've left it with the absolute positioned icons, and using the existing `columns: true` functionality for now as this solution requires the least amount of new code, but let me know if you think I should look at using `flex`.

Apologies if I'm over complicating this...maybe I'm worrying too much about accommodating the `Anything else that might be included that could have quite a long name` example, when it will likely never show up in production.

AshGDS commented:
Thanks @matthillco 👍  - it's a good idea, I did a small spike however and `space-evenly` lead to scenarios like this: 
<img width="824" alt="image" src="https://github.com/user-attachments/assets/5f279616-acab-4ead-8f36-6fdc81bbdc8c">

I appreciate your idea was just a starting point though. My original implementation was a more complex iteration of `display: flex`, but because I had used a `min-width` on the list elements to make the wrapping of items look nicer, and I wasn't using the existing `columns: true` functionality, it was the cause of some of the back and forth in the PR.

I think the problem is trying to make elements of varying widths act and wrap perfectly, and therefore any solution we use is going to have a trade off with either alignment or text wrapping.

I agree that position absolute is fiddly, though as it's the existing functionality of the share links I'm more inclined to keep using it just to make this PR as small as possible. I'm also leaning towards just go with the existing functionality of `columns` for this reason.

matthillco commented:
> Thanks @matthillco 👍 - it's a good idea, I did a small spike however and `space-evenly` lead to scenarios like this: <img alt="image" width="824" src="https://private-user-images.githubusercontent.com/8880610/375724761-5f279616-acab-4ead-8f36-6fdc81bbdc8c.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mjg2NDcxNDIsIm5iZiI6MTcyODY0Njg0MiwicGF0aCI6Ii84ODgwNjEwLzM3NTcyNDc2MS01ZjI3OTYxNi1hY2FiLTRlYWQtOGYzNi02ZmRjODFiYmRjOGMucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MTAxMSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDEwMTFUMTE0MDQyWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ZTM3Zjk0YzMxODdlNjZmMDYzMzdkYjI1NmUyYzlmNmMzOGJiNjhlOGQ2ZTg3OTUzMGIwMTJkZTNmNTI3MDA3YiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.HhFjrTlWYzi-Xt9yxPcxlnO-QDAyj4L0qcceJTMDxYk">
> 
> I appreciate your idea was just a starting point though. My original implementation was a more complex iteration of `display: flex`, but because I had used a `min-width` on the list elements to make the wrapping of items look nicer, and I wasn't using the existing `columns: true` functionality, it was the cause of some of the back and forth in the PR.
> 
> I think the problem is trying to make elements of varying widths act and wrap perfectly, and therefore any solution we use is going to have a trade off with either alignment or text wrapping.
> 
> I agree that position absolute is fiddly, though as it's the existing functionality of the share links I'm more inclined to keep using it just to make this PR as small as possible. I'm also leaning towards just go with the existing functionality of `columns` for this reason.

andysellick commented on app/assets/stylesheets/govuk_publishing_components/components/_share-links.scss:
```diff
@@ -83,3 +83,37 @@ $column-width: 9.5em;
   }
 }
+
+.gem-c-share-links--extra-spacing {
```
I think given that this extra spacing is really only needed for the new icon set I'd suggest combining the two into a single option - I don't think we need them to be separate.

I'd also suggest this because the extra spacing option on the normal icons makes the text align to the bottom of the icons, which I'm not sure looks right.

AshGDS replied:
Thanks @andysellick - I've merged the two down to one now. Have pushed it as a separate commit but will squash eventually.

andysellick reviewed (commented):
I think I'm basically happy with this now, apart from one suggestion.

andysellick reviewed (approved):
Well done - sorry this got so complicated 👏
